### PR TITLE
Sessions: student view results: stop showing a random identifier for anonymous responses #6092

### LINKED ISSUE #6092: Sessions: student view results: stop showing a random identifier for anonymous responses
Currently, all answers by the same student will have the same anonymous identifier e.g., `Anonymous Student 2322423`. That is good to have in some contexts, but some instructors don't want such traceability between responses from the same anonymous respondent. ~~An option to turn it on/off would be nice to have.~~ 
~~EDIT: we can go back to fully randomized anonymous identifiers. Making it an option may be an unnecessary complication.~~
EDIT: we can remove the identifier completely. It's enough to say `Anonymous Student`

### PR DESCRIPTION
**Pull request by bqnguyen94**
Fixes #6092

### CODE REVIEW

**Comment by bqnguyen94** [comment]
On going.

**Comment by bqnguyen94** [comment]
Note to self: Here I limited the changes to `StudentFeedbackResultsPage`, and only in the `PageData` so that they are contained within the view and do not mess with anything else, e.g. csv export, sorted page data, etc. I also didn't modify the `InstructorFeedbackResultsPageData` as anonymity is not crucial for `Instructor` type (?!), and may interfere with building the comment form (need more investigation).

**Comment by bqnguyen94** [comment]
Ready for review.

**Comment by bqnguyen94** [comment]
@damithc I have managed to modify only the generated data for page view to not include the random identifier, but only for `StudentFeedbackResultsPage`. Shall I try the same for `InstructorFeedbackResultsPage` as well?

@reviewer: I have considered removing the random identifier completely from `FeedbackSessionResultsBundle`, but ended up sticking to modifying only the page data to minimize changes to codebase.

**Comment by damithc** [comment]
> @damithc I have managed to modify only the generated data for page view to not include the random identifier, but only for StudentFeedbackResultsPage. Shall I try the same for InstructorFeedbackResultsPage as well?

It will be good to remove it from both sides, especially to prevent instructors identifying students who give anonymous feedback about the instructor/course. But that can be a separate PR and not as important as this one.

**Comment by whipermr5** [comment]
> I have considered removing the random identifier completely from FeedbackSessionResultsBundle, but ended up sticking to modifying only the page data to minimize changes to codebase.

It should be removed completely, if not in this PR then in another.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 1]
```diff
@@ -430,6 +430,17 @@ private static String getHashOfName(String name) {
         return Long.toString(Math.abs((long) name.hashCode()));
     }
 
+    public static String getAnonEmailWithoutHash(FeedbackParticipantType type) {
+        String anonName = getAnonNameWithoutHash(type);
+        return anonName + "@@" + anonName + ".com";
+    }
+
+    public static String getAnonNameWithoutHash(FeedbackParticipantType type) {
```
Comment (whipermr5) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Avoid having references to the hash in new code since the hash is slated to be removed completely from the code base.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 2]
```diff
@@ -430,6 +430,17 @@ private static String getHashOfName(String name) {
         return Long.toString(Math.abs((long) name.hashCode()));
     }
 
+    public static String getAnonEmailWithoutIdentifier(FeedbackParticipantType type) {
```
Comment (whipermr5) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Where is this used?

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 3]
```diff
@@ -430,6 +430,17 @@ private static String getHashOfName(String name) {
...
+        String anonName = getAnonNameWithoutIdentifier(type);
+        return anonName + "@@" + anonName + ".com";
+    }
+
+    public static String getAnonNameWithoutIdentifier(FeedbackParticipantType type) {
```
Comment (whipermr5) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Maybe just `getAnonDisplayedName`?

**Comment by whipermr5** [comment]
Also, need to remove the code that replaces the hash with `participant.hash` for HTML tests: https://github.com/TEAMMATES/teammates/blob/3a330e661879be77ae530efae2d1c6ef2d76bc81/src/test/java/teammates/test/driver/HtmlHelper.java#L437-L439

**Comment by bqnguyen94** [comment]
> Also, need to remove the code that replaces the hash with participant.hash for HTML tests:
As of now, this code is still needed for instructor UI tests as I haven't managed to remove the displayed anonymous hash for instructor views yet.
The best way to do so is to remove hash completely from the code base, and for that I have tried removing hash from `getAnonName` in `FeedbackSessionResultsBundle`, but the updated html test files from `godmode` suggested it's not that simple, in some cases question feedbacks were shown in where they are not intended to (?!), and some were not shown as expected.

File: src/test/resources/pages/studentFeedbackResultsPageNUMSCALE.html [hunk 4]
```diff
@@ -1234,7 +1234,7 @@ <h4>
           <b>
             To:
           </b>
-          Anonymous team ${participant.hash}
+          Anonymous student
```
Comment (whipermr5) [inline src/test/resources/pages/studentFeedbackResultsPageNUMSCALE.html]: Is this an intended effect?

Comment (bqnguyen94) [reply src/test/resources/pages/studentFeedbackResultsPageNUMSCALE.html]: Oh no that is a mistake. In ‘StudentFeedbackResultsPageData:182’ it should have been ‘question.giverType’ instead of ‘PARTICIPANT_TYPE_STUDENT’. There may be more test files affected by this. I will submit a fix tomorrow.

File: src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java [hunk 5]
```diff
@@ -235,4 +240,14 @@ private FeedbackResultsResponseTable createResponseTable(FeedbackQuestionAttribu
         }
         return responsesForRecipient;
     }
+
+    /**
+     * Remove the identifier from anonymous participant name or email, if any.
+     * @param identifier  The given name or email
+     * @return The name or email without any anonymous identifier
+     */
+    private String getDisplayableIdentifier(String identifier) {
```
Comment (whipermr5) [inline src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java]: Oops, I only just realised the approach you were taking (temporarily find and replace the hash). It's okay for the method name to reference the hash then (maybe use `removeAnonymousHash`?) since the method will be removed when the hash is completely removed. But do add a `@deprecated` tag.

File: src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java [hunk 6]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (whipermr5) [inline src/main/java/teammates/common/datatransfer/FeedbackSessionResultsBundle.java]: Unrelated change.

File: src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java [hunk 7]
```diff
@@ -242,11 +242,12 @@ private FeedbackResultsResponseTable createResponseTable(FeedbackQuestionAttribu
...
 
     /**
-     * Remove the identifier from anonymous participant name or email, if any.
+     * Remove the anonymous hash from anonymous participant name or email, if any.
      * @param identifier  The given name or email
-     * @return The name or email without any anonymous identifier
+     * @return The name or email without any anonymous hash
      */
-    private String getDisplayableIdentifier(String identifier) {
+    @Deprecated //Slated to be removed
```
Comment (whipermr5) [inline src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java]: Add a space after the `//`. Maybe elaborate more: `// The anonymous hash is slated for complete removal`

File: src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java [hunk 8]
```diff
@@ -170,22 +173,24 @@ private FeedbackResultsResponseTable createResponseTable(FeedbackQuestionAttribu
...
+                displayedGiverName = removeAnonymousHash(giverName);
             }
 
             boolean isUserRecipient = student.email.equals(response.recipient);
             if (isUserGiver && !isUserRecipient) {
                 // If the giver is the user, show the real name of the recipient
                 // since the giver would know which recipient he/she gave the response to
                 recipientName = bundle.getNameForEmail(response.recipient);
+            } else {
+                recipientName = removeAnonymousHash(recipientName);
```
Comment (whipermr5) [inline src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java]: This can be moved to line 171?

File: src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java [hunk 9]
```diff
@@ -235,4 +238,15 @@ private FeedbackResultsResponseTable createResponseTable(FeedbackQuestionAttribu
         }
         return responsesForRecipient;
     }
+
+    /**
+     * Remove the anonymous hash from anonymous participant name or email, if any.
```
Comment (damithc) [inline src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java]: Removes

Comment (bqnguyen94) [reply src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java]: Done

Comment (whipermr5) [reply src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java]: He meant change `Remove` to `Removes`

Comment (whipermr5) [reply src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java]: I guess removing the javadoc altogether is fine too since this is a temporary method.

Comment (whipermr5) [reply src/main/java/teammates/ui/pagedata/StudentFeedbackResultsPageData.java]: @bqnguyen94 I'll leave it up to you whether you want to add the javadoc back in (but if you do make sure the tense is `Removes` not `Remove`). You can merge your own PR :)
